update-engine/src/display/group_display.rs

```diff
@@ -0,0 +1,478 @@
+                (k, prefix)
+            })
+            .collect();
+        let single_states: BTreeMap<_, _> = keys_and_prefixes
+            .into_iter()
+            .map(|(k, prefix)| (k, SingleState::new(prefix, max_width)))
+            .collect();
+
+        let not_started = single_states.len();
+        Self {
+            writer,
+            max_width,
+            start_instant: None,
+            single_states,
+            formatter: LineDisplayFormatter::new(),
+            stats: GroupDisplayStats::new(not_started),
+        }
+    }
+
+    /// Creates a new `GroupDisplay` with the provided report keys, using the
+    /// `Display` impl to obtain the respective prefixes.
+    pub fn new_with_display(
+        keys: impl IntoIterator<Item = K>,
+        writer: W,
+    ) -> Self
+    where
+        K: fmt::Display,
+    {
+        Self::new(
+            keys.into_iter().map(|k| {
+                let prefix = k.to_string();
+                (k, prefix)
+            }),
+            writer,
+        )
+    }
+
+    /// Sets the styles for all future lines.
+    #[inline]
+    pub fn set_styles(&mut self, styles: LineDisplayStyles) {
+        self.formatter.set_styles(styles);
+    }
+
+    /// Sets the amount of time before new progress events are shown.
+    #[inline]
+    pub fn set_progress_interval(&mut self, interval: Duration) {
+        self.formatter.set_progress_interval(interval);
+    }
+
+    /// Returns true if this `GroupDisplay` is producing reports corresponding
+    /// to the given key.
+    pub fn contains_key<Q>(&self, key: &Q) -> bool
+    where
+        K: Borrow<Q>,
+        Q: Ord,
+    {
+        self.single_states.contains_key(key)
+    }
+
+    /// Adds an event report to the display, keyed by the index, and updates
+    /// internal state.
+    ///
+    /// Returns `Ok(())` if the report was accepted because the key was
+    /// known to this `GroupDisplay`, and an error if it was not.
+    pub fn add_event_report<Q>(
+        &mut self,
+        key: &Q,
+        event_report: EventReport<S>,
+    ) -> Result<(), UnknownReportKey>
+    where
+        K: Borrow<Q>,
+        Q: Ord,
+    {
+        if let Some(state) = self.single_states.get_mut(key) {
+            let result = state.add_event_report(event_report);
+            if self.start_instant.is_none() {
+                self.start_instant = Some(Instant::now());
+            }
+            self.stats.apply_result(result);
+            Ok(())
+        } else {
+            Err(UnknownReportKey {})
+        }
+    }
+
+    /// Writes a "Status" or "Summary" line to the writer with statistics.
+    pub fn write_stats(&mut self, header: &str) -> std::io::Result<()> {
+        // Add a prefix which is equal to the maximum width of the prefixes.
+        // [prefix 00:00:00] takes up self.max_width + 9 characters inside the
+        // brackets.
+        let prefix = " ".repeat(self.max_width);
```

Comment:
Should this be `self.max_width + 9`? If not, I don't follow the comment. :)

Reply:
The code is correct but the comment is confusing. Will rewrite the comment.